File: DSA/DevvratKhandal_590014906/Day_29/Question_1.java
import java.util.*;

class Question_1 {

    public static boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();

        for (char ch : s.toCharArray()) {

            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            } else {

                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();

                if ((ch == ')' && top != '(') ||
                    (ch == '}' && top != '{') ||
                    (ch == ']' && top != '[')) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);

        System.out.println("Enter the bracket string:");
        String s = sc.nextLine();

        boolean result = isValid(s);

        System.out.println("Is the string valid? " + result);

        sc.close();
    }
}

File: DSA/NabhanshRishiGaur_590028370/Day_29/Question_1.c
#include <stdbool.h>
#include <string.h>
#include <stdlib.h>

bool isValid(char * s){

    int n = strlen(s);
    char *stack = (char*)malloc(n);
    int top = -1;

    for(int i = 0; i < n; i++){

        char c = s[i];

        if(c == '(' || c == '[' || c == '{'){
            stack[++top] = c;
        }
        else{

            if(top == -1){
                free(stack);
                return false;
            }

            char t = stack[top--];

            if((c == ')' && t != '(') ||
               (c == ']' && t != '[') ||
               (c == '}' && t != '{')){
                free(stack);
                return false;
            }
        }
    }

    bool valid = (top == -1);
    free(stack);
    return valid;
}
